### ISSUE TITLE
[Due payment 2024-10-03][$250] Thread - Unread marker not displayed when opening thread with unread comments

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#47817]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.23-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y 
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4882295
**Email or phone of affected tester (no customers):**
**Issue reported by:** Applause Internal Team 

## Action Performed:
1. Open the app and log in
2. Navigate to any chat
3. Send a message
4. Start a thread and send some messages
5. Make any comment as unread
6. Navigate to the LHN
7. Open the the thread

## Expected Result:
The new message marker is displayed

## Actual Result:
The new message marker is not displayed

## Workaround:
Unknown

## Platforms:

- [x] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/f7c3e490-dc49-4673-9763-6628a5c6f201

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0166496cbdb268ee61</li>
        <li>Upwork Job ID: 1827933307174227915</li>
        <li>Last Price Increase: 2024-09-02</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @kadiealexander</details>

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The unread marker doesn't show when manually marked as unread and reopen the report. This happens to all report, not limited to thread.

### What is the root cause of that problem?
The logic to display the unread marker checks whether the message is from the current user or not.
https://github.com/Expensify/App/blob/ad7c1d5e394c8a9ee1ce7427233aa085350bb500/src/pages/home/report/ReportActionsList.tsx#L225-L239

If it's from the current user, we won't show the unread marker, but only if the message is not marked manually (`messageManuallyMarkedUnread`).
https://github.com/Expensify/App/blob/ad7c1d5e394c8a9ee1ce7427233aa085350bb500/src/pages/home/report/ReportActionsList.tsx#L172

When we mark it as unread manually, we set the state to the current date.
https://github.com/Expensify/App/blob/ad7c1d5e394c8a9ee1ce7427233aa085350bb500/src/pages/home/report/ReportActionsList.tsx#L256-L258

But when we switch the report or close and reopen it, the new report is mounted, so the state is reset.

### What changes do you think we should make in order to solve the problem?

We need to store the mark manually state outside the component lifecycle.
https://github.com/Expensify/App/blob/ad7c1d5e394c8a9ee1ce7427233aa085350bb500/src/pages/home/report/ReportActionsList.tsx#L135-L136

```
const messageManuallyMarkedUnreadCache = {};
```

Then, initialize the state with the cached value.
https://github.com/Expensify/App/blob/ad7c1d5e394c8a9ee1ce7427233aa085350bb500/src/pages/home/report/ReportActionsList.tsx#L172

```
const [messageManuallyMarkedUnread, _setMessageManuallyMarkedUnread] = useState(messageManuallyMarkedUnreadCache[report.reportID]);
```

Next, we create a function to update the state and cache.
```
const setMessageManuallyMarkedUnread = (isManuallyMarkedUnread: boolean) => {
    _setMessageManuallyMarkedUnread(isManuallyMarkedUnread);
    messageManuallyMarkedUnreadCache[report.reportID] = isManuallyMarkedUnread;
}
```

Last, we need to update the clear code here to clear the cache only.
https://github.com/Expensify/App/blob/ad7c1d5e394c8a9ee1ce7427233aa085350bb500/src/pages/home/report/ReportActionsList.tsx#L214-L219

```
messageManuallyMarkedUnreadCache[report.reportID] = false;
```

_Note: the above solution changes the type from Date to boolean_

**Comment by kadiealexander** [comment]
@mollfpr please review the proposal above!

**Comment by klajdipaja** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The unread marker does not show when the user marks unread a message that they themselves have sent

### What is the root cause of that problem?
The handling of the unread marker depends on the state of messageManuallyMarkedUnread which is lost when the user leaves the screen due to the component itself getting destroyed.

### What changes do you think we should make in order to solve the problem?
When the message is marked as unread a BE action is triggered that updates the Onyx state appropriately. We can rely on that to display the marker line instead of the component state.
To do so we need to change the logic of the shouldDisplayNewMarker  to the one below:

``` javascript

const shouldDisplayNewMarker = (reportAction: OnyxTypes.ReportAction, index: number): boolean => {
            const nextMessage = sortedVisibleReportActions[index + 1];
            const isCurrentMessageUnread = isMessageUnread(reportAction, unreadMarkerTime);
            const isNextMessageRead = !nextMessage || !isMessageUnread(nextMessage, unreadMarkerTime);
            const shouldDisplay = isCurrentMessageUnread && isNextMessageRead && !ReportActionsUtils.shouldHideNewMarker(reportAction);
            const isWithinVisibleThreshold = scrollingVerticalOffset.current < MSG_VISIBLE_THRESHOLD ? reportAction.created < (userActiveSince.current ?? '') : true;
            return shouldDisplay && isWithinVisibleThreshold;
        };
```
Along with it we also need to remove `messageManuallyMarkedUnread` from the component as it becomes unused.

### What alternative solutions did you explore? (Optional)

**Comment by mollfpr** [comment]
Sorry for the delay!

Reviewing 👀

**Comment by mollfpr** [comment]
I am still able to reproduce the issue.

I agree with the RCA in both proposals but @bernhardoj did a great job elaborating on the root cause. I onboard with @bernhardoj [solution](https://github.com/Expensify/App/issues/47817#issuecomment-2303781052) and I think we don't need to use onyx to accommodate our need here.

🎀 👀 🎀 C+ reviewed!

**Comment by klajdipaja** [comment]
@mollfpr Wouldn't it be better to rely on the backend information instead of a state in the local storage so that the information is consistent across different devices? If I as a user mark unread a message in my mobile app I would expect that to show as unread if I open expensify in the browser as well.

**Comment by mollfpr** [comment]
>If I as a user mark unread a message in my mobile app I would expect that to show as unread if I open expensify in the browser as well.

@klajdipaja It's still a big question for me, I tried to trace back the commit and check our test rail case, but I still couldn't figure out what's the expectation on the above case.

The logic itself is added here https://github.com/Expensify/App/pull/45797 and I don't want to remove the logic it may cause a regression.

**Comment by neil-marcellini** [comment]
> I am still able to reproduce the issue.
> 
> I agree with the RCA in both proposals but @bernhardoj did a great job elaborating on the root cause. I onboard with @bernhardoj [solution](https://github.com/Expensify/App/issues/47817#issuecomment-2303781052) and I think we don't need to use onyx to accommodate our need here.
> 
> 🎀 👀 🎀 C+ reviewed!

Thanks for the proposal @bernhardoj. I tried to look into the history like @mollfpr and it's been around for a long time. I don't totally understand why we need to only consider whether the message is from the current user when it's not manually marked as unread.

> @mollfpr Wouldn't it be better to rely on the backend information instead of a state in the local storage so that the information is consistent across different devices? If I as a user mark unread a message in my mobile app I would expect that to show as unread if I open expensify in the browser as well.

That's a very valid point @klajdipaja. Can we please try removing the [isFromCurrentUser check](https://github.com/Expensify/App/blob/ea9b6345afb553c3756e07636d03c517818cd393/src/pages/home/report/ReportActionsList.tsx#L231) and the `!messageManuallyMarkedUnread` condition, or figure out exactly why it's needed? Let's also make this work across devices too. If we need a backend changes that's fine.

**Comment by neil-marcellini** [comment]
Not overdue, waiting for updated proposals

**Comment by klajdipaja** [comment]
@neil-marcellini That's exactly what my proposal does, it removes both conditions. 
In the [PR ](https://github.com/Expensify/App/pull/45797) that this conditions where added the comment above the new condition says: 

>   // Prevent displaying a new marker line when report action is of type "REPORT_PREVIEW" and last actor is the current user

You can found that here: https://github.com/Expensify/App/pull/45797/files#diff-42d45a343b031265784f22ea27e4502b3c1968e492b1fc83154414b584d59217R225

I did test this and I do not see a new marker line. Maybe the marker line was showing up in that specific scenario due to the local state check of `messageManuallyMarkedUnread`.

**Comment by klajdipaja** [comment]
I also found a bug in the existing logic of the unread marker that comes from `isWithinVisibleThreshold` that can be tested by openning a chat with between two users with the following steps:
1. Open chat in user 1
2. Do nothing for a bit, like a minute.
3. Open chat in user 2 and send a message to the chat with user 1
4. Expected result: In user 1 chat the user sees a new message with a new message marker line
Actual result: In user 1 chat the user sees a new message but no marker line

**Comment by neil-marcellini** [comment]
Ok, I would like to go with @klajdipaja's [proposal](https://github.com/Expensify/App/issues/47817#issuecomment-2310040198) instead, because it uses backend data and should work across devices. Let's be sure to test that. Please also test the bugs fixed by messageManuallyMarkedUnread with your PR.

**Comment by neil-marcellini** [comment]
> I also found a bug in the existing logic of the unread marker that comes from `isWithinVisibleThreshold` that can be tested by openning a chat with between two users with the following steps:
> 
> 1. Open chat in user 1
> 2. Do nothing for a bit, like a minute.
> 3. Open chat in user 2 and send a message to the chat with user 1
> 4. Expected result: In user 1 chat the user sees a new message with a new message marker line
>    Actual result: In user 1 chat the user sees a new message but no marker line

Ok interesting. Would you like to fix that in your PR for this issue, or do you think it would be best to keep separate? We should probably do that to keep the PRs small and focused, unless the solution is a small change. If you plan to keep it separate please report the bug in #expensify-bugs and tag me.

**Comment by klajdipaja** [comment]
@neil-marcellini I will work on the PR on Monday. 
I have an idea how to fix the second bug that can fit in the same PR but if that's not the case it would be better to create a separate issue

**Comment by klajdipaja** [comment]
@neil-marcellini Sorry for the delay in creating the PR. It took me more than I thought it would to revive my old MacOs environment for the iOS testing. 
Should I add other reviewers to the PR, I'm thinking @arosiclair since he worked on the original code?

**Comment by arosiclair** [comment]
> Should I add other reviewers to the PR, I'm thinking @arosiclair since he worked on the original code?

Yeah please add me for review I'd like to take a look.

> I also found a bug in the existing logic of the unread marker that comes from isWithinVisibleThreshold that can be tested by openning a chat with between two users with the following steps:
> 
> Open chat in user 1
> Do nothing for a bit, like a minute.
> Open chat in user 2 and send a message to the chat with user 1
> Expected result: In user 1 chat the user sees a new message with a new message marker line
> Actual result: In user 1 chat the user sees a new message but no marker line

I'm pretty sure this is not a bug. If a user has the report open and is scrolled near the bottom (the scroll view is below the threshold - `scrollingVerticalOffset.current < MSG_VISIBLE_THRESHOLD`), then newly received messages should _not_ show the unread marker since the user is already reading them.

**Comment by klajdipaja** [comment]
I am wondering if the process for hiring on upwork has failed or maybe it has changed because the job has not been assigned to me on UpWork as should be based on this description:

```
If your proposal is accepted by the Expensify engineer assigned to the issue, Expensify will hire you on Upwork and assign the GitHub issue to you.

```

**Comment by kadiealexander** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@mollfpr] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@mollfpr] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@mollfpr] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@mollfpr] Determine if we should create a regression test for this bug.
- [ ] [@mollfpr] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@kadiealexander] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by kadiealexander** [comment]
Payouts due:

- [ ] Contributor: $250 @klajdipaja
- [ ] Contributor+: $250 @mollfpr

Upwork job is [here]().

@neil-marcellini is [this](https://github.com/Expensify/App/issues/49751) a regression?

**Comment by klajdipaja** [comment]
@kadiealexander I have accepted the Upwork offer but still has not received the payment. Can you let me know if there was an issue or when I can expect the payment to be made?

**Comment by neil-marcellini** [comment]
@kadiealexander Wow yes this one was a bit tricky. There were some regressions and the PR to fix this issue was mentioned, and maybe revealed some other issues, but it was not reverted so I think we can say that it did not directly cause any regressions. I think it's best explained in [this comment](https://github.com/Expensify/App/issues/49525#issuecomment-2375719614). Let's pay the full amount.

We can discuss more if anyone passionately disagrees.

**Comment by kadiealexander** [comment]
Sounds great to me, thanks @neil-marcellini!

**Comment by mollfpr** [comment]
> [@mollfpr] The PR that introduced the bug has been identified. Link to the PR:
 [@mollfpr] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

I couldn't determined which PR causing the issue.

> [@mollfpr] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

I think the regression step is enough.

> [@mollfpr] Determine if we should create a regression test for this bug.
 [@mollfpr] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.

**Mark unread user messages**

1. Open a chat with a workspace or another user
2. Send a few messages in the chat
3. Mark a message as unread
4. Open a different chat
5. Open the same chat again
6. _**Verify**_ the "new message" marker line is shown on the message that was marked unread

**Unread messages in the thread**

1. Open a chat with a workspace or another user
2. Send a message in the chat
3. Start a new thread in the chat and send a few messages
4. Mark one of the messages as unread
5. Leave the thread
6. Open the thread again
7. _**Verify**_ the "new message" marker line is shown on the message that was marked unread


**Unread messages between devices**

1. Run all the steps for one of the previous tests in the desktop app
2. Open a different device (mweb, android, ios)
3. Open the same chat where you marked a message unread
4. _**Verify**_ the "new message" marker line is shown on the message that was marked unread

**Comment by garrettmknight** [comment]
$250 approved for @mollfpr
